### PR TITLE
Re #32: until with Runnable version that allows throwing exception

### PR DESCRIPTION
**Pull request by tkowalcz**

### CODE REVIEW

File: awaitility/src/main/java/com/jayway/awaitility/core/AssertionCondition.java [hunk 1]
```diff
@@ -118,3 +137,21 @@ public Void await() {
         return null;
     }
 }
+
+class ThrowingRunnableAdapter implements ThrowingRunnable {
+
+    private final Runnable supplier;
+
+    public ThrowingRunnableAdapter(Runnable supplier) {
+        this.supplier = supplier;
```
Comment (szpak) [inline awaitility/src/main/java/com/jayway/awaitility/core/AssertionCondition.java]: An assertion that supplier cannot me null?

**Comment by szpak** [comment]
Changes in production code look good to me.

It's not the issue of this PR, but there is a lot of duplicated code in tests (which is clearly visible here).

In `AwaitilityJava8Test` all tests are duplicated, but I don't see any other solution that using parameterized Spock tests with method name passes as a String. Unfortunately it is a hack itself (`await()."$methodName"(...)`) and it is fragile in case of recatoring, so even I don't think it is worth to do it :(.

Another thing is `ConditionEvaluationListenerJava8Test` where conditionEvaluationListener inline code is repeated several times. By moving `lastMatchMessage` and `countDown` to fields I was able to extract `ConditionEvaluationListener` to a field as well and the class was reduced by ~150 lines. As those tests verifies an error message only the `ConditionEvaluationListener` itself is a minor thing and the visibility of intentions in given tests could be considered as better. WDYT about that guys?

```
private AtomicReference<String> lastMatchMessage = new AtomicReference<>();
private CountDown countDown = new CountDown(10);

private ConditionEvaluationListener conditionEvaluationListener = condition -> {
                    try {
                        countDown.call();
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                    lastMatchMessage.set(condition.getDescription());
                };

@Test(timeout = 2000)
public void expectedMatchMessageForAssertionConditionsWhenUsingLambdasWithoutAlias() {
    with()
            .conditionEvaluationListener(conditionEvaluationListener)
            .until(() -> assertEquals(5, (int) countDown.get()));

    String expectedMatchMessage = String.format("%s reached its end value", ConditionEvaluationListenerJava8Test.class.getName());
    assertThat(lastMatchMessage.get(), allOf(startsWith("Condition defined as a lambda expression"), endsWith(expectedMatchMessage)));
}

(...)
```

(a small caveat is that one more tests is needed to be written to check lamba parameters if a local variable is used in a lambda condition)

**Comment by tkowalcz** [comment]
I did the refactoring of `ConditionEvaluationListenerJava8Test` along the lines you suggested and agree that it looks much better and is quite clear. 

One other small thing is that I would change `CountDown::get` to return `int` instead of `Integer` so the cast to int in `(int) countDown.get()` goes away.

**Comment by johanhaleby** [comment]
So let me summary to see if I've understood things correctly:

1. You want to be able to throw a checked exception from the lambda expression passed to until?
1. Your pull request deals with this by introducing the ThrowingRunnable?
1. Unfortunately we cannot name the method taking an instance of ThrowingRunnable `until` since it will clash with `Runnable` so thus you've created a new method called `untilPass`. 

Is this correct?

**Comment by tkowalcz** [comment]
This sums it pretty nicely.

**Comment by johanhaleby** [comment]
I can see why this would be a useful feature. Another option (but this is probably a bad idea) is to replace to until method we have today that takes `Runnable` and replace it with `ThrowingRunnable`. Lambda expressions would still work but legacy code would need to migrate. For example if you previously did like this:

```java
Runnable condition = new Runnable() { .. } 
await().until(condition);
```

we could force legacy users to transform the `Runnable` into a `ThrowingRunnable`, for example like this:

```java
Runnable condition = new Runnable() { .. } 
await().until(matches(condition));
```

where `matches` may simply be a static method somewhere defined like this:

```java
public ThrowingRunnable matches(Runnable r) {
   return new ThrowingRunnable() {
           public void run() throws Exception {
               r.run();
           }
   }
}
```

But as I said I doubt that this would be the best solution.. But I'm still not convinced that we should add an additional method to the API (like `untilPass`).

**Comment by tkowalcz** [comment]
I understand your reluctance, unfortunately so far I was unable to figure out other way.

**Comment by johanhaleby** [comment]
Would this it be an option to somewhere include this method:

```java
public static Callable<Boolean> matches(ThrowingRunnable throwingRunnable) {
    return new Callable<Boolean>() {
        @Override
        public Boolean call() throws Exception {
            try {
                throwingRunnable.run();
            } catch (Throwable e) {
                return false;
            }
            return true;
        }
    };
}
```

which you can then use like this:

```java
await().until(matches(() -> someMethodThatThrowsCheckedException("test"));
```

What do you think?

**Comment by szpak** [comment]
With just wrapping `ThrowingRunnable` with `Callable<Boolean>` we would probably lose a nice error message from AssertJ.

File: awaitility/src/main/java/com/jayway/awaitility/core/AssertionCondition.java [hunk 2]
```diff
@@ -118,3 +137,25 @@ public Void await() {
...
+
+    public ThrowingRunnableAdapter(Runnable supplier) {
+        if (supplier == null) {
+            throw new IllegalArgumentException("You must specify a supplier (was null).");
+        }
+
+        this.supplier = supplier;
+    }
+
+    @Override
```
Comment (szpak) [inline awaitility/src/main/java/com/jayway/awaitility/core/AssertionCondition.java]: Btw, Awaitility is compiled in Java 5 compatibility code. `@Override` for interfaces was introduced in Java 6.

**Comment by johanhaleby** [comment]
You're right @szpak. How about this:

```java
public static class SafeExceptionRethrower {

	public static void safeRethrow(Throwable t) {
		SafeExceptionRethrower.<RuntimeException> safeRethrow0(t);
	}

	@SuppressWarnings("unchecked")
	private static <T extends Throwable> void safeRethrow0(Throwable t) throws T {
		throw (T) t;
	}
}
```

```java
public static Runnable matches(ThrowingRunnable throwingRunnable) {
    return new Runnable() {
        @Override
        public void run() {
            try {
                throwingRunnable.run();
            } catch (Throwable e) {
                SafeExceptionRethrower.safeRethrow(e);
            }
        }
    };
}
```

Usage:

```java
await().until(matches(() -> someMethodThatThrowsCheckedException("test"));
```

It's a bit dirty but I suppose it workds and `SafeExceptionRethrower` has been removed from Awaitility but I can add it again (as an internal class).

**Comment by szpak** [comment]
It's tricky, but I have to agree that it seems to work and no new method is needed to be added (not to mention that the implementation is much easier). As a drawback it would not be easy for a user to find that method (that approach) without reading JavaDoc, but having multiple methds can be confusing as well. @tkowalcz WDYT?

Btw, probably you just wanted to remove [that class](https://github.com/jayway/awaitility/blob/master/awaitility/src/main/java/com/jayway/awaitility/core/SafeExceptionRethrower.java) ;)

**Comment by johanhaleby** [comment]
Oh so it's still there :) Well that's makes it even easier to implement if we decide to go this route.

**Comment by johanhaleby** [comment]
WDYT?

**Comment by johanhaleby** [comment]
I may release a new version of Awaitilty today and since I haven't heard anything I'll go with my suggestion and add a wrapper (something similar to `matches`). If you don't agree *please* speak up ASAP!

**Comment by tkowalcz** [comment]
Hi, I am swamped recently, sorry for that! I am ok with this approach.

File: awaitility/src/main/java/com/jayway/awaitility/core/AssertionCondition.java (hunk 2 above)
Comment (tkowalcz) [inline awaitility/src/main/java/com/jayway/awaitility/core/AssertionCondition.java]: I don't have Java 5 JDK so was a bit cheating, but hey - the CI passed :)

**Comment by johanhaleby** [comment]
@tkowalcz Thanks for your comments. Just one quick thing if you can find the time: Should the method be called `matches` or should it be called something else? Perhaps `throwing` or something similar would be better?

**Comment by johanhaleby** [comment]
I've named it "throwing". Please tell me (soon) if this was a bad idea :)

**Comment by tkowalcz** [comment]
I'd prefer 'matches' or 'success'/'succeedes' since ability to throw exception is only a convenience and the main purpose of that call remains the same (to evaluate a bunch of asserts).

**Comment by johanhaleby** [comment]
Alright, then I'll rename it to matches. Thanks!

**Comment by tkowalcz** [comment]
Thank you!

**Pull request closed without merging**